[Bug]: New metadata filtering feature in 0.8.29 for simple vector store now prevents deleting documents

### Bug Description

It looks like a bug has been introduced in #7564, as part of the 0.8.29 release.

When calling `index.delete_ref_doc(doc_id, delete_from_docstore=True)`, 

From inspecting my code it looks like my `SimpleVectorStore._data` has a new attribute `metadata_dict` which is empty. So when running  through this loop:
```
        for text_id in text_ids_to_delete:
            del self._data.embedding_dict[text_id]
            del self._data.metadata_dict[text_id]
            del self._data.text_id_to_ref_doc_id[text_id]
```

I get a KeyError. 

I notice in the `query` method there is some safety built in around the new metadata filters: 

```
    def query(
        self,
        query: VectorStoreQuery,
        **kwargs: Any,
    ) -> VectorStoreQueryResult:
        """Get nodes for response."""
        # Prevent metadata filtering on stores that were persisted without metadata.
        if (
            query.filters is not None
            and self._data.embedding_dict
            and not self._data.metadata_dict
        ):
            raise ValueError(
                "Cannot filter stores that were persisted without metadata. "
                "Please rebuild the store with metadata to enable filtering."
            )
```

Can I propose a similar safety is added to the deletion of source nodes for indexes that were persisted without metadata. I shouldn't have to rebuild my index in order to delete documents. 

### Version

0.8.29

### Steps to Reproduce

- Load a Simple Vector Store index from llama-index version < 0.8.29 which contains documents.
- Call `index.delete_ref_doc(doc_id, delete_from_docstore=True)`

### Relevant Logs/Tracbacks

```shell
delete_ref_doc
    self._vector_store.delete(ref_doc_id)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <llama_index.vector_stores.simple.SimpleVectorStore object at 0x122296610>
ref_doc_id = 'bcbec082-a406-4de1-beca-58c06ad22106', delete_kwargs = {}
text_ids_to_delete = {'9d75f807-2345-4ecd-9f01-1fdf2419e490'}
text_id = '9d75f807-2345-4ecd-9f01-1fdf2419e490'
ref_doc_id_ = 'cbc4b033-e78d-42bd-a331-8e55cf082fc1'

    def delete(self, ref_doc_id: str, **delete_kwargs: Any) -> None:
        """
        Delete nodes using with ref_doc_id.
    
        Args:
            ref_doc_id (str): The doc_id of the document to delete.
    
        """
        text_ids_to_delete = set()
        for text_id, ref_doc_id_ in self._data.text_id_to_ref_doc_id.items():
            if ref_doc_id == ref_doc_id_:
                text_ids_to_delete.add(text_id)
    
        for text_id in text_ids_to_delete:
            del self._data.embedding_dict[text_id]
>           del self._data.metadata_dict[text_id]
E           KeyError: '9d75f807-2345-4ecd-9f01-1fdf2419e490'
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -9,6 +9,7 @@
 - Normalize scores returned from ElasticSearch vector store (#7792)
 - Fixed `refresh_ref_docs()` bug with order of operations (#7664)
 - Delay postgresql connection for `PGVectorStore` until actually needed (#7793)
+- Fix KeyError in delete method of `SimpleVectorStore` related to metadata filters (#7829)
 
 ## [0.8.33] - 2023-09-25
 

--- a/llama_index/vector_stores/simple.py
+++ b/llama_index/vector_stores/simple.py
@@ -156,8 +156,11 @@ class SimpleVectorStore(VectorStore):
 
         for text_id in text_ids_to_delete:
             del self._data.embedding_dict[text_id]
-            del self._data.metadata_dict[text_id]
             del self._data.text_id_to_ref_doc_id[text_id]
+            # Handle metadata_dict not being present in stores that were persisted without metadata.
+            # Or not being present for nodes stored prior to metadata functionality.
+            if self._data.metadata_dict is not None:
+                self._data.metadata_dict.pop(text_id, None)
 
     def query(
         self,

--- a/llama_index/vector_stores/simple.py
+++ b/llama_index/vector_stores/simple.py
@@ -157,8 +157,9 @@ class SimpleVectorStore(VectorStore):
         for text_id in text_ids_to_delete:
             del self._data.embedding_dict[text_id]
             del self._data.text_id_to_ref_doc_id[text_id]
-            # Handle metadata_dict not being present in stores that were persisted without metadata.
-            # Or not being present for nodes stored prior to metadata functionality.
+            # Handle metadata_dict not being present in stores that were persisted
+            # without metadata, or, not being present for nodes stored
+            # prior to metadata functionality.
             if self._data.metadata_dict is not None:
                 self._data.metadata_dict.pop(text_id, None)
 

